history/history.c

```diff
@@ -0,0 +1,1079 @@
+    check_rc(rc);
+
+    installed_map = (char *)calloc(map_size, sizeof(char));
+    check_ptr(installed_map);
+
+    rc = db_play_transaction(ctx->db, trans_id, installed_map, map_size);
+    check_rc(rc);
+
+    history_set_ids_from_map(ctx, installed_map, map_size);
+
+error:
+    safe_free(installed_map);
+
+    return rc;
+}
+
+/* Update state in ctx and db by setting a baseline state.
+ * This is used to initialize the db, but can be called later to set a new baseline.
+ */
+int history_record_state(struct history_ctx *ctx)
+{
+    int rc = 0;
+    char *err_msg = NULL;
+    int trans_id = 0;
+
+    /* avoid unfinished transaction record on failure or crash */
+    rc = sqlite3_exec(ctx->db, "BEGIN TRANSACTION;", 0, 0, NULL);
+    check_db_rc(ctx->db, rc);
+
+    rc = sqlite3_exec(ctx->db,
+        "CREATE TABLE IF NOT EXISTS "
+            "transactions(Id INTEGER PRIMARY KEY AUTOINCREMENT,"
+                "cookie TEXT, cmdline TEXT,"
+                "timestamp INTEGER, "
+                "type INTEGER);",
+        0, 0, NULL);
+    check_db_rc(ctx->db, rc);
+
+    rc = db_add_transaction(ctx->db, &trans_id, "(set)", time(NULL),
+                            ctx->cookie, HISTORY_TRANS_TYPE_BASE);
+    check_rc(rc);
+
+    rc = sqlite3_exec(ctx->db,
+        "CREATE TABLE IF NOT EXISTS "
+            "trans_items(Id INTEGER PRIMARY KEY AUTOINCREMENT,"
+            "trans_id INTEGER,"
+            "type INTEGER,"
+            "rpm_id INTEGER);",
+        0, 0, &err_msg);
+    check_db_rc(ctx->db, rc);
+
+    rc = db_add_trans_items(ctx->db, trans_id, HISTORY_ITEM_TYPE_SET,
+                            ctx->installed_ids, ctx->installed_count);
+    check_rc(rc);
+
+error:
+    if (rc)
+        sqlite3_exec(ctx->db, "ROLLBACK;", 0, 0, NULL);
+    else
+        sqlite3_exec(ctx->db, "COMMIT;", 0, 0, NULL);
+    return rc;
+}
+
+void history_free_transactions(struct history_transaction *tas, int count)
+{
+    if (tas) {
+        for(int i = 0; i < count; i++) {
+            safe_free(tas[i].cookie);
+            safe_free(tas[i].cmdline);
+            safe_free(tas[i].delta.added_ids);
+            safe_free(tas[i].delta.removed_ids);
+        }
+        free(tas);
+    }
+}
+
+/*
+   Get a list of transcactions and store them into buffer pointed to by ptas.
+   The number will be stored into pcount. Optionally a range can be set with
+   the from and to parameters. Either both have to be set to a range, or both
+   need to be 0 for no range (all transactions). The list will be in reverse
+   order if the reverse parameter is set.
+*/
+int history_get_transactions(struct history_ctx *ctx,
+                             struct history_transaction **ptas,
+                             int *pcount,
+                             int reverse, int from, int to)
+{
+    sqlite3_stmt *res = NULL;
+    int step, rc = 0;
+    int i, count = 0;
+    struct history_transaction *tas = NULL;
+    char sql[256];
+
+    check_cond(to >= from);
```

Comment:
Redundant check before calling this function we already check From > To

Reply:
Yes, it's redundant in this case, but maybe future changes add callers that omit the check. This is an API function where we should do some checking, but if this was just a helper function we could skip it.